### PR TITLE
Add portable report

### PR DESCRIPTION
**Pull request by donaldcampbelljr**
Allow setting portable flag for `pipestat summarize` such that report resources (images, pdfs) are copied to a resource folder within the report folder, allowing for the entire HTML report to be more easily shared across systems.

Example CLI usage:
`pipestat summarize -c /home/drc/pepatac_tutorial/processed/looper_pipestat_config.yaml --portable True`

### CODE REVIEW

File: docs/changelog.md [hunk 1]
```diff
@@ -2,6 +2,9 @@
 
 This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html) and [Keep a Changelog](https://keepachangelog.com/en/1.0.0/) format.
 
+## [0.7.1] - 2024-xx-xx
+### Added
+- added `portable` flag to allow for easier sharing of reports.
```
Comment (khoroshevskyi) [inline docs/changelog.md:7]:
```suggestion
- added `portable` flag to allow easier sharing of reports. 
```

File: pipestat/reports.py [hunk 2]
```diff
@@ -40,21 +41,30 @@
 class HTMLReportBuilder(object):
     """Generate HTML summary report for project/samples"""
 
-    def __init__(self, prj):
+    def __init__(self, prj, portable):
```
Comment (khoroshevskyi) [inline pipestat/reports.py:44]: I think you want to add default value here

File: tests/test_pipestat.py [hunk 3]
```diff
@@ -1126,6 +1126,39 @@ def test_basics_project_html_report(
             htmlreportpath = psm.summarize(amendment="")
             assert htmlreportpath is not None
 
+    @pytest.mark.parametrize("backend", ["file", "db"])
```
Comment (khoroshevskyi) [inline tests/test_pipestat.py:1129]: Nice!

File: tests/test_pipestat.py [hunk 4]
```diff
@@ -2095,6 +2128,38 @@ def test_pipestat_iter(
 
 
 class TestMultiResultFiles:
+    # @pytest.mark.parametrize("backend", ["file"])
```
Comment (khoroshevskyi) [inline tests/test_pipestat.py:2131]: If you won't use this code, and everything works fine, delete this code before merge

File: pipestat/reports.py [hunk 5]
```diff
@@ -100,6 +110,32 @@ def __call__(
         self.create_index_html(navbar, self.create_footer())
         return self.index_html_path
 
+    def _create_copy_for_porting(self, parentpath, record_identifier):
```
Comment (khoroshevskyi) [inline pipestat/reports.py:113]: add typing

File: pipestat/reports.py [hunk 6]
```diff
@@ -100,6 +110,32 @@ def __call__(
...
+
+        :param parentpath: path of existing file
+        :param newpath: desired destination
+        :return: newpath if successful else returns parentpath
+        """
+        destination_dir = os.path.join(self.pipeline_reports, "resources")
+        full_name = os.path.basename(parentpath)
+
+        destinationfile = os.path.join(destination_dir, record_identifier)
+        destinationfile = os.path.join(destinationfile, full_name)
```
Comment (khoroshevskyi) [inline pipestat/reports.py:127]: you can join more then 2 files at the same time

**Review by khoroshevskyi — APPROVED** [review]
I didn't test actual functionality, but LGTM
Just few small changes needed

File: docs/changelog.md (hunk 1 above)
Comment (donaldcampbelljr) [reply docs/changelog.md:7]: This is tricky. I believe 'allow for' is actually what we want here, but I'm happy to see evidence to the contrary.

But I think I'm going to change this as:
`added portable flag to pipestat summarize to create a shareable version of the html report`

File: pipestat/reports.py (hunk 2 above)
Comment (donaldcampbelljr) [reply pipestat/reports.py:44]: I had it set below, but I moved it to the parameter level as you suggested.

File: tests/test_pipestat.py (hunk 4 above)
Comment (donaldcampbelljr) [reply tests/test_pipestat.py:2131]: Done

File: pipestat/reports.py (hunk 5 above)
Comment (donaldcampbelljr) [reply pipestat/reports.py:113]: done

File: pipestat/argparser.py [hunk 7]
```diff
@@ -334,6 +334,13 @@ def add_subparser(
             help="project or sample level pipeline type. ",
         )
 
+        sps[cmd].add_argument(
+            "--portable",
+            type=bool,
+            metavar="X",
```
Comment (nsheff) [inline pipestat/argparser.py:340]: why is the metavar X?

Comment (nsheff) [reply pipestat/argparser.py:340]: is that typical for boolean args?

File: pipestat/reports.py (hunk 6 above)
Comment (nsheff) [reply pipestat/reports.py:127]: yes, this should be one line

File: pipestat/reports.py [hunk 8]
```diff
@@ -368,6 +412,21 @@ def create_object_htmls(self, navbar, footer):
                         pass
                     else:
                         try:
+                            if self.portable:
+                                newimagepath = self._create_copy_for_porting(
```
Comment (nsheff) [inline pipestat/reports.py:416]: I would suggest using underscores to separate words in variable names, like `new_image_path` -- that's the style we're using elsewhere in these packages.

File: pipestat/reports.py [hunk 9]
```diff
@@ -368,6 +412,21 @@ def create_object_htmls(self, navbar, footer):
...
+                                    parentpath=sample_result[image_result]["path"],
```
Comment (nsheff) [inline pipestat/reports.py:417]:
```suggestion
                                    parent_path=sample_result[image_result]["path"],
```

File: pipestat/reports.py [hunk 10]
```diff
@@ -368,6 +412,21 @@ def create_object_htmls(self, navbar, footer):
...
+                                    record_identifier=sample_name,
+                                )
+
+                                sample_result[image_result]["path"] = newimagepath
```
Comment (nsheff) [inline pipestat/reports.py:421]:
```suggestion
                                sample_result[image_result]["path"] = new_image_path
```

File: pipestat/reports.py [hunk 11]
```diff
@@ -368,6 +412,21 @@ def create_object_htmls(self, navbar, footer):
...
+
+                                newthumbnailpath = self._create_copy_for_porting(
```
Comment (nsheff) [inline pipestat/reports.py:423]:
```suggestion
                                new_thumbnail_path = self._create_copy_for_porting(
```

File: pipestat/reports.py [hunk 12]
```diff
@@ -368,6 +412,21 @@ def create_object_htmls(self, navbar, footer):
...
+                                    parentpath=sample_result[image_result]["thumbnail_path"],
+                                    record_identifier=sample_name,
+                                )
+
+                                sample_result[image_result]["thumbnail_path"] = newthumbnailpath
```
Comment (nsheff) [inline pipestat/reports.py:428]:
```suggestion
                                sample_result[image_result]["thumbnail_path"] = new_thumbnail_path
```

**Comment by nsheff** [comment]
Should this go ahead and zip up the portable report?

File: pipestat/reports.py (hunk 6 above)
Comment (donaldcampbelljr) [reply pipestat/reports.py]: done

**Comment by donaldcampbelljr** [comment]
> Should this go ahead and zip up the portable report?

Yeah, I was thinking that would be a nice addition.

File: pipestat/reports.py (hunk 8 above)
Comment (donaldcampbelljr) [reply pipestat/reports.py]: Done!

**Comment by donaldcampbelljr** [comment]
I've changed the CLI such that you do not have to input a bool:
`pipestat summarize -c ./looper_pipestat_config.yaml --portable
`

**Comment by donaldcampbelljr** [comment]
Closes #139

**Pull request merged**